subsys/bluetooth/mesh/dfu_slot.c

```diff
@@ -147,145 +139,176 @@ bt_mesh_dfu_slot_add(size_t size, const uint8_t *fwid,
 		return NULL;
 	}
 
-	slot->slot.fwid_len = fwid_len;
-	slot->slot.metadata_len = metadata_len;
-	slot->slot.uri_len = uri_len;
-	memcpy(slot->slot.fwid, fwid, fwid_len);
-	memcpy(slot->slot.metadata, metadata, metadata_len);
-	memcpy(slot->slot.uri, uri, uri_len);
-	slot->slot.size = size;
-
-	err = slot_store(slot);
-	if (err) {
-		slot_invalidate(slot);
-		LOG_WRN("Store failed (err: %d)", err);
-		return NULL;
+	if (slot_index == UINT32_MAX) {
+		slot_index_defrag();
 	}
 
-	sys_slist_append(&list, &slot->n);
+	slot->slot.fwid_len = 0;
+	slot->slot.metadata_len = 0;
+	slot->slot.size = 0;
+	slot->idx = ++slot_index;
+
+	LOG_DBG("Reserved slot #%u", slot - &slots[0]);
 
-	LOG_DBG("Added slot #%u: %s", slot - &slots[0],
-		bt_hex(slot->slot.fwid, slot->slot.fwid_len));
 	return &slot->slot;
 }
 
-int bt_mesh_dfu_slot_valid_set(const struct bt_mesh_dfu_slot *slot, bool valid)
+int bt_mesh_dfu_slot_info_set(struct bt_mesh_dfu_slot *dfu_slot, size_t size,
+			      const uint8_t *metadata, size_t metadata_len)
 {
-	uint16_t idx;
-	bool prev;
-	int err;
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
+
+	if (metadata_len > CONFIG_BT_MESH_DFU_METADATA_MAXLEN) {
+		return -EFBIG;
+	}
+
+	if (slot->idx == 0 || is_slot_committed(slot)) {
+		return -EINVAL;
+	}
+
+	slot->slot.size = size;
+	slot->slot.metadata_len = metadata_len;
+	memcpy(slot->slot.metadata, metadata, metadata_len);
+	return 0;
+}
 
-	if (!SLOT_IN_ARRAY(slot) || !slot_in_use(slot)) {
-		return -ENOENT;
+int bt_mesh_dfu_slot_fwid_set(struct bt_mesh_dfu_slot *dfu_slot,
+			      const uint8_t *fwid, size_t fwid_len)
+{
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
+
+	if (fwid_len > CONFIG_BT_MESH_DFU_FWID_MAXLEN) {
+		return -EFBIG;
+	}
+
+	if (slot->idx == 0 || is_slot_committed(slot)) {
+		return -EINVAL;
 	}
 
-	idx = slot_idx(slot);
+	for (int i = 0; i < ARRAY_SIZE(slots); i++) {
+		if (slots[i].idx != 0 &&
+		    slot_eq(&slots[i].slot, fwid, fwid_len)) {
+			return -EALREADY;
+		}
+	}
```

Comment:
What I had in place was the "allow uploading the same firmware, but once one is stored, the other one will be dropped after the upload (in bt_mesh_dfu_slot_commit)."-solution you mention here.

This was change after this comment from you:

> It seems it is already too late to check this. `_commit` is called after a fw is received which is usually a long process. We should not allow starting uploading two identical firmware ids. We will also ideally need to check this error code in dfd_srv.c, but currently we just ignore it:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/702ec88e3c9da13932b6780ab0ef3624408bc344/subsys/bluetooth/mesh/dfd_srv.c#L828
> 
> 
> From what I see uniqueness of the fwid is already checked in dfd_srv.c in `bt_mesh_dfd_srv_oob_check_complete` and in `handle_upload_start`, so it looks redundant. Since dfu slot api is internal, it should be safe to remove this check.

Do you want me to change it back?

Reply:
Right, then the only option is to prevent starting to upload the same firmware.